src/socket.rs

```diff
@@ -68,8 +68,25 @@ use crate::{MaybeUninitSlice, RecvFlags};
 /// # Ok(()) }
 /// ```
 pub struct Socket {
-    // The `sys` module most have access to the socket.
-    pub(crate) inner: sys::Socket,
+    inner: Inner,
+}
+
+/// Store a `TcpStream` internally to take advantage of its niche optimizations on Unix platforms.
+pub(crate) type Inner = std::net::TcpStream;
```

Comment:
Does `File` has the same optimisation? Perhaps using `TcpStream` is misleading...

Reply:
Can you remove the `Inner` type and create `Socket` directly in the `sys::socket_from_raw` module directly? Or is there a specific reason you need the type?

Reply:
Windows can't use a `File` because that stores a handle instead of a socket. And I think it's better to use `TcpStream` on both Windows and Unix, because it's possible that in the future Windows will get this niche optimization as well (since -1 is not a valid socket value I don't think).

The reason I used `Inner` was so that `TcpStream` could only be stated once in the code, where we can explain what it does. If someone sees:

```rust
pub(crate) fn socket_from_raw(socket: Socket) -> crate::Socket {
    crate::Socket { inner: unsafe { TcpStream::from_raw_fd(socket) } }
}
```

that's slightly more confusing since there's a random unexplained `TcpStream` being created in a place quite far from `Socket`'s declaration and the corresponding explanation. But I don't really mind, I'm happy to change it.

Reply:
👍 clear.